albatross/core/distribution.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_CORE_DISTRIBUTION_H
+#define ALBATROSS_CORE_DISTRIBUTION_H
+
+#include <vector>
+#include <iostream>
+#include <Eigen/Core>
+#include "indexing.h"
+
+namespace albatross {
+
+/*
+ * A Distribution holds what is typically assumed to be a
+ * multivariate Gaussian distribution with mean and optional
+ * covariance.
+ */
+template <typename CovarianceType>
+struct Distribution {
+  Eigen::VectorXd mean;
+  CovarianceType covariance;
+
+  std::size_t size() const {
+    // If the covariance is defined it must have the same number
+    // of rows and columns which should be the same size as the mean.
+    assert_valid();
+    return mean.size();
+  }
+
+  void assert_valid() const {
+    if (covariance.size() > 0) {
```

Comment:
Could we make `CovarianceType` optional?

Reply:
I was trying to weigh the benefit of this approach versus including another submodule.  Think it's worth adding https://github.com/akrzemi1/Optional ?

Reply:
Yeah, probably not necessary at the moment, but something to think about in the future.

Reply:
yeah I agree.  If we run into another situation where `optional` would be nice I think it'd be worth adding it (and changing this).